feat: Twitter-like hexagon profile images for NFT avatars

svitekpavel opened the pull request:
## What does this PR do?

TODO
- [x] Figure out how to distinguish between NFT avatar and simple image
- [x] Review all places where avatar image is used

Fixes #50

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)
## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- Review the displayed avatars in all changed components

## Screenshots

<img width="574" alt="Screenshot 2022-11-18 at 14 02 18" src="https://user-images.githubusercontent.com/988706/202711416-da7a136b-bb6b-4dd9-b67d-c5ce82b7a310.png">
<img width="506" alt="Screenshot 2022-11-18 at 14 02 04" src="https://user-images.githubusercontent.com/988706/202711419-8a5844e8-8845-451a-a321-5b74ab876309.png">
<img width="448" alt="Screenshot 2022-11-18 at 14 01 53" src="https://user-images.githubusercontent.com/988706/202711421-308d8372-e6ed-42dd-ab63-af9f1c3f6f62.png">
<img width="275" alt="Screenshot 2022-11-18 at 14 01 48" src="https://user-images.githubusercontent.com/988706/202711428-c0a41ff7-4ba3-49c2-beb4-00dcbb8fd7c4.png">
<img width="871" alt="Screenshot 2022-11-18 at 14 01 45" src="https://user-images.githubusercontent.com/988706/202711430-6c86a146-fb4a-458a-a11e-0e31b6aca862.png">

shanpriyan commented on src/components/Profile/Details.tsx:
```diff
@@ -63,19 +63,56 @@ const Details: FC<Props> = ({ profile }) => {
-  return (
-    <div className="px-5 mb-4 space-y-5 sm:px-0">
+  const UserAvatarNFT = () => (
+    <div>
       <div className="relative -mt-24 w-32 h-32 sm:-mt-32 sm:w-52 sm:h-52">
         <img
           src={getAvatar(profile)}
           className="w-32 h-32 bg-gray-200 rounded-xl ring-8 ring-gray-50 sm:w-52 sm:h-52 dark:bg-gray-700 dark:ring-black"
+          style={{
+            // @ts-ignore
+            'clip-path': 'url("#hex-hw-shapeclip-clipconfig")',
+            'background-image': `url(${getAvatar(profile)})`,
+            'background-position': 'center',
+            'background-size': 'cover',
+            'background-repeat': 'no-repeat'
+          }}
           height={128}
           width={128}
           alt={profile?.handle}
         />
       </div>
+      <svg height="0" viewBox="0 0 400 376" width="0">
+        <defs>
+          <clipPath
+            clipPathUnits="objectBoundingBox"
+            id="hex-hw-shapeclip-clipconfig"
+            transform="scale(0.005 0.005319148936170213)"
+          >
+            <path d="M193.248 69.51C185.95 54.1634 177.44 39.4234 167.798 25.43L164.688 20.96C160.859 15.4049 155.841 10.7724 149.998 7.3994C144.155 4.02636 137.633 1.99743 130.908 1.46004L125.448 1.02004C108.508 -0.340012 91.4873 -0.340012 74.5479 1.02004L69.0879 1.46004C62.3625 1.99743 55.8413 4.02636 49.9981 7.3994C44.155 10.7724 39.1367 15.4049 35.3079 20.96L32.1979 25.47C22.5561 39.4634 14.0 ...
+          </clipPath>
+        </defs>
+      </svg>
+    </div>
```
```suggestion
    <>
      <div className="relative -mt-24 w-32 h-32 sm:-mt-32 sm:w-52 sm:h-52">
        <img
          src={getAvatar(profile)}
          className="w-32 h-32 bg-gray-200 rounded-xl ring-8 ring-gray-50 sm:w-52 sm:h-52 dark:bg-gray-700 dark:ring-black"
          style={{
            // @ts-ignore
            'clip-path': 'url("#hex-hw-shapeclip-clipconfig")',
            'background-image': `url(${getAvatar(profile)})`,
            'background-position': 'center',
            'background-size': 'cover',
            'background-repeat': 'no-repeat'
          }}
          height={128}
          width={128}
          alt={profile?.handle}
        />
      </div>
      <svg height="0" viewBox="0 0 400 376" width="0">
        <defs>
          <clipPath
            clipPathUnits="objectBoundingBox"
            id="hex-hw-shapeclip-clipconfig"
            transform="scale(0.005 0.005319148936170213)"
          >
            <path d="M193.248 69.51C185.95 54.1634 177.44 39.4234 167.798 25.43L164.688 20.96C160.859 15.4049 155.841 10.7724 149.998 7.3994C144.155 4.02636 137.633 1.99743 130.908 1.46004L125.448 1.02004C108.508 -0.340012 91.4873 -0.340012 74.5479 1.02004L69.0879 1.46004C62.3625 1.99743 55.8413 4.02636 49.9981 7.3994C44.155 10.7724 39.1367 15.4049 35.3079 20.96L32.1979 25.47C22.5561 39.4634 14.0458 54.2034 6.74789 69.55L4.39789 74.49C1.50233 80.5829 0 87.2441 0 93.99C0 100.736 1.50233 107.397 4.39789 113.49L6.74789 118.43C14.0458 133.777 22.5561 148.517 32.1979 162.51L35.3079 167.02C39.1367 172.575 44.155 177.208 49.9981 180.581C55.8413 183.954 62.3625 185.983 69.0879 186.52L74.5479 186.96C91.4873 188.32 108.508 188.32 125.448 186.96L130.908 186.52C137.638 185.976 144.163 183.938 150.006 180.554C155.85 177.17 160.865 172.526 164.688 166.96L167.798 162.45C177.44 148.457 185.95 133.717 193.248 118.37L195.598 113.43C198.493 107.337 199.996 100.676 199.996 93.93C199.996 87.1841 198.493 80.5229 195.598 74.43L193.248 69.51Z" />
          </clipPath>
        </defs>
      </svg>
    </>
```

bigint commented:
This should be supported only for users who are using NFT avatars 🙇🏼

svitekpavel commented:
> This should be supported only for users who are using NFT avatars 🙇🏼

@bigint Yes, that's the idea. This PR is still work in progress. My intention was also to show that I'm working on this, so that other people don't start working on this in parallel.

svitekpavel commented:
@bigint I updated all components which I think should display NFT avatar as hexagon. There are few components where I didn't change the default avatar in circle as I'm not sure if we want to change them. Examples: `ProfileCircles`, `MessageList`, `Messages/Preview`, `Profile/MutualFollower`.

I updated the screenshots in the PR description.

Can you please review the remaining places which I listed above if we want to use hexagon avatars there as well?

svitekpavel commented:
> @bigint I updated all components which I think should display NFT avatar as hexagon. There are few components where I didn't change the default avatar in circle as I'm not sure if we want to change them. Examples: `ProfileCircles`, `MessageList`, `Messages/Preview`, `Profile/MutualFollower`.
> 
> I updated the screenshots in the PR description.
> 
> Can you please review the remaining places which I listed above if we want to use hexagon avatars there as well?

@bigint this is synced as well. Please also have a look at my notes when you are reviewing this PR. Thanks

bigint commented:
@svitekpavel Can we also support borders for NFT avatars?

RN in profile page the big avatar looks weird without border 🙇🏼

bigint commented on apps/web/src/components/UI/Avatar.tsx:
```diff
@@ -0,0 +1,62 @@
+import { type Profile } from 'lens';
+import React, { type DetailedHTMLProps, type FC, type ImgHTMLAttributes, type ReactElement } from 'react';
+
+type ImgProps = DetailedHTMLProps<ImgHTMLAttributes<HTMLImageElement>, HTMLImageElement>;
+type Props = ImgProps & {
+  profile?: Profile | null;
+  isNft?: boolean;
+  render?: (imgProps: ImgProps) => ReactElement;
+};
```
**suggestion:** we always use interface everywhere 🙏🏼

bigint commented:
> @svitekpavel Can we also support borders for NFT avatars?
> 
> RN in profile page the big avatar looks weird without border 🙇🏼

@svitekpavel 🙇🏼
